### PR TITLE
Dt 1669

### PR DESCRIPTION
**Pull request by vesameskanen**
All known history/page transition bugs fixed, including:

- In desktop, browsing through itineraries no longer pushes new entries to history
- Rerouting at summary page resets selected route index back to first (=best) (fixes DT-1651)
- In mobile, expanding/collapsing the map in the stop view no longer pushes into history
- Route page / tab switch between pysäkit/aikatalu/häiriöt no longer pushes into history
- In mobile, using Lahellasi/Suosikit pushes tab opening into history once (no switches between the tabs though, to avoid broblems in getting back to index page). Back button works now as expected.
- Search modal opened from the splash screen pushes to history. Back button works now as expected.
- Search modal opened from the route summary page pushes to history.
- Search modal opened from the edit favourite page pushes to history.
- Disruption info modal now closes from the back button.
- SearchMainContainer opened from IndexPage (search field / origin popup) pushes to history

### CODE REVIEW

File: app/component/AppBarLarge.js [hunk 1]
```diff
@@ -2,35 +2,47 @@ import React, { PropTypes } from 'react';
...
+  );
+};
 
 AppBarLarge.propTypes = {
   titleClicked: PropTypes.func.isRequired,
 };
 
 AppBarLarge.contextTypes = {
-  executeAction: React.PropTypes.func.isRequired,
+  router: React.PropTypes.object,
```
Comment (hannesj) [inline app/component/AppBarLarge.js]: can we use https://github.com/ReactTraining/react-router/blob/v2.8.1/docs/API.md#proptypes for these?

File: app/component/OneTabSearchModal.js [hunk 2]
```diff
@@ -52,15 +56,19 @@ class OneTabSearchModal extends React.Component {
...
       });
     }
-
-    return this.props.closeModal();
+    this.context.router.goBack();
   };
 
+  modalIsOpen = () => (
+    this.context.location.state ?
+      !(!this.context.location.state.oneTabSearchModalOpen) : false
```
Comment (hannesj) [inline app/component/OneTabSearchModal.js]: !(! ?

File: app/component/SearchMainContainer.js [hunk 3]
```diff
@@ -13,44 +14,55 @@ import SearchModalLarge from './SearchModalLarge';
...
-
-  componentDidMount() {
-    this.context.getStore('SearchStore').addChangeListener(this.onSearchChange);
-  }
-
-  componentWillUnmount() {
-    this.context.getStore('SearchStore').removeChangeListener(this.onSearchChange);
+  componentDidUpdate() {
+    if (this.props.searchModalIsOpen) {
+      setTimeout(
```
Comment (hannesj) [inline app/component/SearchMainContainer.js]: do we need a setTimeout for some reason?

File: app/component/SearchMainContainer.js [hunk 4]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (hannesj) [inline app/component/SearchMainContainer.js]: extra newline?

File: app/component/Splash.js [hunk 5]
```diff
@@ -9,28 +9,28 @@ import Intro from './Intro';
...
+      state: {
+        ...this.context.location.state,
+        oneTabSearchModalOpen: true,
+      },
     });
   };
 
   renderContents() {
+    const modalOpen =
+      !!(this.context.location.state && this.context.location.state.oneTabSearchModalOpen);
```
Comment (hannesj) [inline app/component/Splash.js]: !! converts it to a boolean, is that necessary?

File: app/component/IndexPage.js [hunk 6]
```diff
@@ -136,30 +135,63 @@ class IndexPage extends React.Component {
     }
   }
 
-  openFavourites = () => {
-    this.replace('/suosikit');
+  push = (path) => {
```
Comment (hannesj) [inline app/component/IndexPage.js]: is this needed for some reason?

**Comment by hannesj** [comment]
Should we write some tests for these changes in behaviour?

File: app/component/OneTabSearchModal.js (hunk 2 above)
Comment (vesameskanen) [inline app/component/OneTabSearchModal.js]: parens apparently not needed for boolean casting

File: app/component/SearchMainContainer.js (hunk 3 above)
Comment (vesameskanen) [inline app/component/SearchMainContainer.js]: Did not actually test, the same trick was just used before. I will retest if it works fine without settimeout.

File: app/component/SearchMainContainer.js (hunk 4 above)
Comment (vesameskanen) [inline app/component/SearchMainContainer.js]: Oops ... fixed.

File: app/component/IndexPage.js (hunk 6 above)
Comment (vesameskanen) [inline app/component/IndexPage.js]: The replace param? I am not sure. Anyway, if you switch between the tabs a couple of times in mobile,  without the additional logic you will have  to click equally many times to get back to the front page.  Now back takes you there immediately. That feels convenient but I may be wrong.

**Comment by vesameskanen** [comment]
Many bugs were fixed so regression tests might make sense. I guess there is no test which checks that  browser back button works.

File: app/component/Splash.js (hunk 5 above)
Comment (vesameskanen) [inline app/component/Splash.js]: Sub component expect boolean. Without cast, console shows warnings.

Comment (hannesj) [inline app/component/Splash.js]: What about using new Boolean() for clarity?

File: app/component/SearchMainContainer.js (hunk 3 above)
Comment (hannesj) [inline app/component/SearchMainContainer.js]: Great!

File: app/component/IndexPage.js (hunk 6 above)
Comment (hannesj) [inline app/component/IndexPage.js]: I was thinking about the wrappers

**Comment by hannesj** [comment]
https://github.com/HSLdevcom/digitransit-ui/blob/7f4d8ed370a206974d8e2ea287c9dac392c4ba10/test/flow/tests/favourite/add router's is a test using the back button

**Comment by vesameskanen** [comment]
Thanks .... I will create some new tests.

File: app/component/Splash.js (hunk 5 above)
Comment (vesameskanen) [inline app/component/Splash.js]: Agreed.

File: app/component/AddFavouriteContainer.js [hunk 7]
```diff
@@ -18,6 +18,7 @@ class AddFavouriteContainer extends React.Component {
     intl: intlShape.isRequired,
     executeAction: PropTypes.func.isRequired,
     router: PropTypes.object.isRequired,
+    location: PropTypes.object.isRequired,
```
Comment (hannesj) [inline app/component/AddFavouriteContainer.js]: can we use https://github.com/ReactTraining/react-router/blob/v2.8.1/docs/API.md#proptypes for these?

**Pull request closed without merging**